docs/style-spec/expressions.md

```diff
@@ -0,0 +1,174 @@
+
+###Constants:
+- `[ "ln2" ] -> Number`
+- `[ "pi" ] -> Number`
+- `[ "e" ] -> Number`
+
+### Type conversion:
+- `["string", e: Value] -> String`
+- `["number", e:Value] -> Number`
+  - Uses platform-default string-to-number conversion. (TBD: parse locale-specificformatted number strings)
+- `["boolean", e:Value] -> Boolean`
+  - `0`, `''`, and `null` are converted to `false`, all other values to `true`
+
+###Lookup:
+- `["get", obj: JSONObject, key: String ] -> Value`
+- `["has", obj: JSONObject, key: String ] -> Boolean`
+- `["at", arr: JSONArray, index: Number] -> Value`
+- `["at", arr: Array<T>|Vector<T>, index: Number] -> T`
+- `["typeof", expr: Value] -> String`
+- `["length", e: JSONArray|Vector<T>|String] -> Number`
+
+###Feature data:
+- `["properties"] -> JSONObject` the feature's `properties` object
+- `["geometry_type"] -> String` the string value of `feature.geometry.type`
+- `[ "id" ] -> Value` returns the value of `feature.id`.
+
+###Decision:
+- `["case", cond1: Boolean, result1: T, cond2: Boolean, result2: T, ..., cond_m, result_m: T, result_otherwise: T] -> T`
+- `["match", x: T, a_1: T, y_1: U, a_2: T, y_2: U, ..., a_m: T, y_m: U, y_else: U]` - `a_1`, `a_2`, ... must be _literal_ values of type `T`.
+- `["is_error", expr: T]` - `true` if `expr` is an `Error` value, `false` otherwise
+
+###Comparison and boolean operations:
+- `[ "==", expr1: T, expr2: T] -> Boolean`, where T is any primitive type. (similar for `!=`)
+- `[ ">", lhs_expr: T, rhs_expr: T ] -> Boolean`, where T is any primitive type. (similar for <, >=, <=)
+- `[ "&&", e1: Boolean, e2: Boolean, ... ] -> Boolean` (similar for `||`)
+- `[ "!", e: Boolean] -> Boolean`
+
+###Curves:
+
+`["curve", interpolation, x: Number, n_1: Number, y_1: T, ..., n_m: Number, y_m: T] -> T` defines a function with `(n_1, y_1)`, ..., `(n_m, y_m)` as input/output pairs, and `interpolation` dictating how inputs between `n_i` and `n_(i+1)` are computed.
+- The `n_i`'s must be numeric literals in strictly ascending order (`n_1 < n_2 < n_3 < ...`)
+- Specific `interpolation` types may imply certain restrictions on the output type `T`.
+- `interpolation` is one of:
+  * `["step"]` - equivalent to existing "interval" function behavior.
+  * `["exponential", base]` - `base` is a number > 0; equivalent to existing "exponential" function behavior. `T` must be `Number` or `Color`
+  * `["linear"]` - equivalent to `["exponential", 1]`
+  * `["cubic-bezier", x1, y1, x2, y2]` - define your own interpolation. `T` must be `Number`
+
+###Math:
+All of the following take `Number` inputs and produce a `Number`.
+- +, -, \*, /, %, ^ (e.g. `["+", expr1, expr2, expr3, …]`, `["-", expr1, expr2 ]`, etc.)
+- log10, ln, log2
+- sin, cos, tan, asin, acos, atan
+- ceil, floor, round, abs
+- min, max
+
+###String:
+- `["concat", expr1: T, expr2: U, …] -> String`
```

Comment:
Would there by any advantage to having a string formatting function instead of concat?

On one hand it would add complexity because it would add function-specific syntax. On the other hand you could handle the entire "this is what I want the displayed text to be" in a single function, which might be more UI-friendly.